state/secrets.go

```diff
@@ -552,14 +554,191 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
 	})
 }
 
-// GrantSecret implements SecretsService.
-func (st *State) GrantSecret(uri *secrets.URI, scope, subject names.Tag, role secrets.SecretRole) error {
-	return errors.NotImplementedf("GrantSecret")
+type secretPermissionDoc struct {
+	DocID string `bson:"_id"`
+
+	Scope   string `bson:"scope-tag"`
+	Subject string `bson:"subject-tag"`
+	Role    string `bson:"role"`
 }
 
-// RevokeSecret implements SecretsService.
-func (st *State) RevokeSecret(uri *secrets.URI, scope, subject names.Tag) error {
-	return errors.NotImplementedf("RevokeSecret")
+func (st *State) findSecretAccessScopeEntity(tag names.Tag) (entity Lifer, collName, docID string, err error) {
+	id := tag.Id()
+	switch tag.(type) {
+	case names.RelationTag:
+		entity, err = st.KeyRelation(id)
+		collName = relationsC
+		docID = id
+	case names.UnitTag:
+		entity, err = st.Unit(id)
+		collName = unitsC
+		docID = id
+	case names.ApplicationTag:
+		entity, err = st.Application(id)
+		collName = applicationsC
+		docID = id
+	default:
+		err = errors.NotValidf("secret scope reference %q", tag.String())
+	}
+	return entity, collName, docID, err
+}
+
+// GrantSecretAccess saves the secret access role for the subject with the specified scope.
+func (st *State) GrantSecretAccess(uri *secrets.URI, scope names.Tag, subject names.Tag, role secrets.SecretRole) (err error) {
+	if role == "" || !role.IsValid() {
```

Comment:
If we think `RoleNone` is invalid then we just need `if !role.IsValid()`

Reply:
it's not valid to grant role none, that is done using revoke.
so it's ok to disallow role none in grant operation
changed "" to RoleNone